File: App/Posts/PostViewModel.m
//  PostViewModel.m
//
//  Copyright 2013 Awful Contributors. CC BY-NC-SA 3.0 US https://github.com/Awful/Awful.app

#import "PostViewModel.h"
#import "AwfulHTMLRendering.h"
#import "AwfulSettings.h"
#import "Awful-Swift.h"

@implementation PostViewModel

- (instancetype)initWithPost:(Post *)post
{
    self = [super init];
    if (!self) return nil;
    
    _post = post;
    
    return self;
}

- (instancetype)init
{
    NSAssert(nil, @"Use -initWithPost: instead");
    return [self initWithPost:nil];
}

- (NSString *)HTMLContents
{
    NSString *innerHTML = self.post.innerHTML;
    if (!innerHTML) return nil;
    HTMLDocument *document = [HTMLDocument documentWithString:innerHTML];
    RemoveSpoilerStylingAndEvents(document);
    RemoveEmptyEditedByParagraphs(document);
    UseHTML5VimeoPlayer(document);
    HighlightQuotesOfPostsByUserNamed(document, [AwfulSettings sharedSettings].username);
    ProcessImgTags(document, ![AwfulSettings sharedSettings].showImages);
    if (self.post.ignored) {
        MarkRevealIgnoredPostLink(document);
    }
    return [document firstNodeMatchingSelector:@"body"].innerHTML;
}

static void MarkRevealIgnoredPostLink(HTMLDocument *document)
{
    HTMLElement *link = [document firstNodeMatchingSelector:@"a[title=\"DON'T DO IT!!\"]"];
    if (!link) return;
    NSURLComponents *components = [NSURLComponents componentsWithString:link[@"href"]];
    components.fragment = @"awful-ignored";
    link[@"href"] = components.URL.absoluteString;
}

- (NSURL *)visibleAvatarURL
{
    return [self showAvatars] ? self.post.author.avatarURL : nil;
}

- (NSURL *)hiddenAvatarURL
{
    return [self showAvatars] ? nil : self.post.author.avatarURL;
}

- (BOOL)showAvatars
{
    return [AwfulSettings sharedSettings].showAvatars;
}

- (NSString *)roles
{
    User *author = self.post.author;
    NSMutableString *roles = [author.authorClasses mutableCopy];
    if ([author isEqual:self.post.thread.author]) {
        [roles appendString:@" op"];
    }
	return roles;
}

- (NSString *)accessibilityRoles
{
    NSArray *unsortedRoles = [self.roles componentsSeparatedByCharactersInSet:[NSCharacterSet whitespaceAndNewlineCharacterSet]];
    NSMutableArray *accessibilityRoles = [unsortedRoles mutableCopy];
    const NSString * const rolesToSpokenRoles[] = {
        @"ik", @"internet knight",
        @"op", @"original poster",
    };
    for (NSUInteger i = 0, end = sizeof(rolesToSpokenRoles) / sizeof(*rolesToSpokenRoles); i < end; i += 2) {
        NSUInteger roleIndex = [accessibilityRoles indexOfObject:rolesToSpokenRoles[i]];
        if (roleIndex != NSNotFound) {
            [accessibilityRoles replaceObjectAtIndex:roleIndex withObject:rolesToSpokenRoles[i + 1]];
        }
    }
    return [accessibilityRoles componentsJoinedByString:@"; "];
}

- (BOOL)authorIsOP
{
    return [self.post.author isEqual:self.post.thread.author];
}

- (NSDateFormatter *)postDateFormat
{
    return [NSDateFormatter postDateFormatter];
}

- (NSDateFormatter *)regDateFormat
{
    return [NSDateFormatter regDateFormatter];
}

- (id)valueForUndefinedKey:(NSString *)key
{
    return [self.post valueForKey:key];
}

@dynamic author;
@dynamic beenSeen;
@dynamic postDate;
@dynamic postID;

@end

File: App/Posts/PostViewModel.swift
//  PostViewModel.swift
//
//  Copyright 2016 Awful Contributors. CC BY-NC-SA 3.0 US https://github.com/Awful/Awful.app

import AwfulCore
import Foundation
import HTMLReader
import Mustache

struct PostViewModel: MustacheBoxable {
    private let dict: [String: Any]

    init(_ post: Post) {
        var roles: String {
            guard let author = post.author else { return "" }
            var roles = author.authorClasses ?? ""
            if post.thread?.author == author {
                roles += " op"
            }
            return roles
        }
        var accessibilityRoles: String {
            let spokenRoles = [
                "ik": "internet knight",
                "op": "original poster",
                ]
            return roles
                .components(separatedBy: .whitespacesAndNewlines)
                .map { spokenRoles[$0] ?? $0 }
                .joined(separator: "; ")
        }
        var showAvatars: Bool {
            return AwfulSettings.shared().showAvatars
        }
        var hiddenAvatarURL: URL? {
            return showAvatars ? nil : post.author?.avatarURL
        }
        var htmlContents: String {
            return massageHTML(post.innerHTML ?? "", isIgnored: post.ignored)
        }
        var visibleAvatarURL: URL? {
            return showAvatars ? post.author?.avatarURL : nil
        }

        dict = [
            "accessibilityRoles": accessibilityRoles,
            "author": [
                "regdate": post.author?.regdate as Any,
                "userID": post.author?.userID as Any,
                "username": post.author?.username as Any],
            "beenSeen": post.beenSeen,
            "hiddenAvatarURL": hiddenAvatarURL as Any,
            "htmlContents": htmlContents,
            "postDate": post.postDate as Any,
            "postID": post.postID,
            "roles": roles,
            "showAvatars": showAvatars,
            "visibleAvatarURL": visibleAvatarURL as Any]
    }
    
    init(author: User, isOP: Bool, postDate: Date, postHTML: String) {
        dict = [
            "author": [
                "regdate": author.regdate as Any,
                "userID": author.userID,
                "username": author.username as Any],
            "beenSeen": false,
            "hiddenAvatarURL": (showAvatars ? author.avatarURL : nil) as Any,
            "htmlContents": massageHTML(postHTML, isIgnored: false),
            "postDate": postDate,
            "postID": "fake",
            "roles": (isOP ? "op " : "") + (author.authorClasses ?? ""),
            "showAvatars": showAvatars,
            "visibleAvatarURL": (showAvatars ? author.avatarURL : nil) as Any]
    }

    var mustacheBox: MustacheBox {
        return Box(dict)
    }
}

private func massageHTML(_ html: String, isIgnored: Bool) -> String {
    let document = HTMLDocument(string: html)
    document.removeSpoilerStylingAndEvents()
    document.removeEmptyEditedByParagraphs()
    document.addAttributeToTweetLinks()
    document.useHTML5VimeoPlayer()
    document.highlightQuotesOfPosts(byUserNamed: AwfulSettings.shared().username)
    document.processImgTags(shouldLinkifyNonSmilies: !AwfulSettings.shared().showImages)
    if !AwfulSettings.shared().autoplayGIFs {
        document.stopGIFAutoplay()
    }
    if isIgnored {
        document.markRevealIgnoredPostLink()
    }
    return document.bodyElement?.innerHTML ?? ""
}

private var showAvatars: Bool {
    return AwfulSettings.shared().showAvatars
}

private extension HTMLDocument {
    func markRevealIgnoredPostLink() {
        guard
            let link = firstNode(matchingSelector: "a[title=\"DON'T DO IT!!\"]"),
            let href = link["href"],
            var components = URLComponents(string: href)
            else { return }
        components.fragment = "awful-ignored"
        guard let replacement = components.url?.absoluteString else { return }
        link["href"] = replacement
    }
}
